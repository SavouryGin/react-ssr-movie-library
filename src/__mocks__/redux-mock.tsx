import { combineReducers } from '@reduxjs/toolkit';
import { moviesInitialState, moviesSlice } from 'store/movies/slice';
import { movieList } from './movie-list';

export const mockedReducer = combineReducers({
  movies: moviesSlice.reducer,
});

export const mockedState = {
  movies: { ...moviesInitialState, list: movieList },
};
