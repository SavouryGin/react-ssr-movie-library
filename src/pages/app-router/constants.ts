export const ROOT_PATH = '/';
export const SEARCH_PATH = '/search';
