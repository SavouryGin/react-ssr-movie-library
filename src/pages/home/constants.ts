import { Genre } from 'enums/genre';

export const GENRE_TABS: Genre[] = [Genre.Drama, Genre.Comedy, Genre.Animation, Genre.Adventure, Genre.Family, Genre.Horror];
