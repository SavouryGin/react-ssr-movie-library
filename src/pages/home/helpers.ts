import { Genre } from 'enums/genre';
import { MovieItem } from 'types/movies';

export const genreChecker = {
  isDocumentary: (item: MovieItem): boolean => {
    const genreList = item.genres.map((option) => option.value);

    return genreList.includes(Genre.Biography) || genreList.includes(Genre.Documentary);
  },
  isComedy: (item: MovieItem): boolean => {
    const genreList = item.genres.map((option) => option.value);

    return genreList.includes(Genre.Comedy);
  },
  isHorror: (item: MovieItem): boolean => {
    const genreList = item.genres.map((option) => option.value);

    return genreList.includes(Genre.Horror);
  },
  isCrime: (item: MovieItem): boolean => {
    const genreList = item.genres.map((option) => option.value);

    return genreList.includes(Genre.Crime);
  },
  isAction: (item: MovieItem): boolean => {
    const genreList = item.genres.map((option) => option.value);

    return genreList.includes(Genre.Action);
  },
};
