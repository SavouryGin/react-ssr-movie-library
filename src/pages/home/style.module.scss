@import 'styles/sizes.scss';

.home {
  padding: $horizontal-padding $vertical-padding;
  position: relative;
}
