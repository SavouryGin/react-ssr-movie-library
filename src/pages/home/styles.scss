@import 'styles/sizes.scss';

.home-page {
  padding: $horizontal-padding $vertical-padding;
}
