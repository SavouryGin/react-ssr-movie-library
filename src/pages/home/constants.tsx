import MoviePanel from 'components/movie-panel';
import React from 'react';
import { TabItem } from 'types/tabs';
import { genreChecker } from './helpers';
import { movieList } from '__mocks__/movie-list';

export const searchFormInitialValue = { movie: '' };

const documentaryMovies = movieList.filter((item) => genreChecker.isDocumentary(item));
const comedyMovies = movieList.filter((item) => genreChecker.isComedy(item));
const horrorMovies = movieList.filter((item) => genreChecker.isHorror(item));
const crimeMovies = movieList.filter((item) => genreChecker.isCrime(item));
const actionMovies = movieList.filter((item) => genreChecker.isAction(item));

export const homeTabs: TabItem[] = [
  {
    tabContent: <MoviePanel items={movieList} />,
    tabTitle: 'All',
    tabId: 'all-movies',
  },
  {
    tabContent: <MoviePanel items={documentaryMovies} />,
    tabTitle: 'Documentary',
    tabId: 'documentary-movies',
  },
  {
    tabContent: <MoviePanel items={comedyMovies} />,
    tabTitle: 'Comedy',
    tabId: 'comedy-movies',
  },
  {
    tabContent: <MoviePanel items={horrorMovies} />,
    tabTitle: 'Horror',
    tabId: 'horror-movies',
  },
  {
    tabContent: <MoviePanel items={crimeMovies} />,
    tabTitle: 'Crime',
    tabId: 'crime-movies',
  },
  {
    tabContent: <MoviePanel items={actionMovies} />,
    tabTitle: 'Action',
    tabId: 'action-movies',
  },
];
