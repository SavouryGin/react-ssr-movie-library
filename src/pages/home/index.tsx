import MoviePanel from 'components/movie-panel';
import React from 'react';
import TabList from 'components/tab-list';
import { TabItem } from 'types/tabs';
import { movieList } from '__mocks__/movie-list';
import './styles.scss';

const Home = () => {
  const homeTabs: TabItem[] = [
    {
      tabContent: <MoviePanel items={movieList} />,
      tabTitle: 'All',
      tabId: 'all-movies',
    },
    {
      tabContent: <>Test 2</>,
      tabTitle: 'Documentary',
      tabId: 'documentary-movies',
    },
    {
      tabContent: <>Test 3</>,
      tabTitle: 'Comedy',
      tabId: 'comedy-movies',
    },
    {
      tabContent: <>Test 4</>,
      tabTitle: 'Horror',
      tabId: 'horror-movies',
    },
    {
      tabContent: <>Test 5</>,
      tabTitle: 'Crime',
      tabId: 'crime-movies',
    },
  ];

  return (
    <article className='home-page'>
      <TabList tabs={homeTabs} className='home-page__tabs' />
    </article>
  );
};

export default Home;
