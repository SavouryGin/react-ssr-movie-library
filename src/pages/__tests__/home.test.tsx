import Home from 'pages/home';
import React from 'react';
import { GENRE_TABS } from 'pages/home/constants';
import { mockedReducer, mockedState } from '__mocks__/redux-mock';
import { renderWithRedux } from '__mocks__/test-utils';
import { screen } from '@testing-library/react';

describe('Home page:', () => {
  it('matches the snapshot', () => {
    const { asFragment } = renderWithRedux(<Home />, mockedReducer, mockedState);
    expect(asFragment()).toMatchSnapshot();
  });

  it('contains the genre tabs with correct ids', () => {
    renderWithRedux(<Home />, mockedReducer, mockedState);
    GENRE_TABS.forEach((tabName) => {
      const id = `${tabName}-movies`;
      const tab = screen.getByTestId(id);
      expect(tab).toBeInTheDocument();
      expect(tab).toHaveProperty('id', id);
    });
  });

  it('contains the search movie form', () => {
    renderWithRedux(<Home />, mockedReducer, mockedState);
    expect(screen.getByRole('form')).toBeInTheDocument();
    expect(screen.getByText('Find your movie')).toBeInTheDocument();
  });
});
