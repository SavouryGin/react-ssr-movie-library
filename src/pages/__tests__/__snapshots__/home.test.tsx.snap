// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page: matches the snapshot 1`] = `
<DocumentFragment>
  <article>
    <div>
      <h2>
        Find your movie
      </h2>
      <form
        aria-label="search movie form"
      >
        <div
          class=""
        >
          <input
            autocomplete="off"
            id="text_input_movie"
            name="movie"
            placeholder="What do you want to watch?"
            type="text"
            value=""
          />
        </div>
        <div>
          <button
            class=""
            id="search-movie-button"
            type="submit"
          >
            <span
              class=""
            >
              Submit
            </span>
          </button>
        </div>
      </form>
    </div>
    <div
      class=""
    >
      <div
        role="tablist"
      >
        <h3
          aria-selected="true"
          class="undefined"
          data-testid="all-movies"
          id="all-movies"
          role="tab"
        >
          All
        </h3>
        <h3
          aria-selected="false"
          class=""
          data-testid="Drama-movies"
          id="Drama-movies"
          role="tab"
        >
          Drama
        </h3>
        <h3
          aria-selected="false"
          class=""
          data-testid="Comedy-movies"
          id="Comedy-movies"
          role="tab"
        >
          Comedy
        </h3>
        <h3
          aria-selected="false"
          class=""
          data-testid="Animation-movies"
          id="Animation-movies"
          role="tab"
        >
          Animation
        </h3>
        <h3
          aria-selected="false"
          class=""
          data-testid="Adventure-movies"
          id="Adventure-movies"
          role="tab"
        >
          Adventure
        </h3>
        <h3
          aria-selected="false"
          class=""
          data-testid="Family-movies"
          id="Family-movies"
          role="tab"
        >
          Family
        </h3>
        <h3
          aria-selected="false"
          class=""
          data-testid="Horror-movies"
          id="Horror-movies"
          role="tab"
        >
          Horror
        </h3>
      </div>
      <section
        aria-labelledby="all-movies"
        role="tabpanel"
      >
        <div
          class=""
          data-cy="movie-panel"
        >
          <div
            class=""
          >
            <div
              style="width: 64px; height: 64px;"
            />
          </div>
          <div
            class=""
          >
            <label
              for="select_sortMovies"
            >
              Sort by
            </label>
            <select
              id="select_sortMovies"
              label="Sort by"
              name="sortMovies"
            >
              <option
                value="0"
              >
                Most popular
              </option>
              <option
                value="2"
              >
                Release (newest)
              </option>
              <option
                value="1"
              >
                Release (oldest)
              </option>
              <option
                value="3"
              >
                Title (A - Z)
              </option>
              <option
                value="4"
              >
                Title (Z - A)
              </option>
              <option
                value="6"
              >
                Rating (highest)
              </option>
              <option
                value="5"
              >
                Rating (lowest)
              </option>
              <option
                value="8"
              >
                Runtime (longest)
              </option>
              <option
                value="7"
              >
                Runtime (shortest)
              </option>
            </select>
          </div>
          <p>
            <strong>
              0
            </strong>
             movies found
          </p>
          <ul />
        </div>
      </section>
    </div>
  </article>
</DocumentFragment>
`;
